### ISSUE TITLE
Move getMatchesProperty into createAdapter for Ripple

### ISSUE THREAD

**Issue by BerndWessels** [material-components/material-components-web#469]
### What MDC-Web Version are you using?

`0.8.0`

### What is the expected behavior?

Rendering `material-components-web` components on the server-side `node.js` should work.

### What is the actual behavior?

Rendering on the server `node.js` results in the following error:

`ReferenceError: HTMLElement is not defined`

### Any other information you believe would be useful?

Tested with [this example](https://github.com/material-components/material-components-web/blob/master/framework-examples/react/src/Checkbox.js).

Also tested with a `preact` version of the same example:

```js
/**
 * Import dependencies.
 */
import {h, Component} from 'preact';
import classnames from 'classnames/dedupe';

import {getCorrectEventName} from '@material/animation/dist/mdc.animation';
import {MDCRipple, MDCRippleFoundation} from '@material/ripple/dist/mdc.ripple';
import {MDCCheckboxFoundation} from '@material/checkbox/dist/mdc.checkbox';
import R from 'ramda';

/**
 * Import local dependencies.
 */

/**
 * Import styles.
 */
import styles from './styles';
import '@material/checkbox/mdc-checkbox.scss';
import '@material/ripple/mdc-ripple.scss';

/**
 * Create the component.
 *
 * static propTypes = {
 * id: PropTypes.string,
 * labelId: PropTypes.string,
 * checked: PropTypes.bool,
 * disabled: PropTypes.bool,
 * indeterminate: PropTypes.bool,
 * onChange: PropTypes.func
 * }
 * static defaultProps = {
 * checked: false,
 * disabled: false,
 * indeterminate: false,
 * onChange: () => {}
 * }
 */
// function getMatchesProperty(HTMLElementPrototype) {
//   return [
//     'webkitMatchesSelector', 'msMatchesSelector', 'matches',
//   ].filter((p) => p in HTMLElementPrototype).pop();
// }
//
// const {ANIM_END_EVENT_NAME} = MDCCheckboxFoundation.strings;
//
// const MATCHES = getMatchesProperty(HTMLElement.prototype);
const MATCHES = true;
export default class Checkbox extends Component {

  /**
   * Initialize local component state.
   */
  constructor(props) {
    super(props);
    this.refs = {};
    this.state = {
      classes: [],
      rippleCss: {},
      checkedInternal: this.props.checked || false,
      disabledInternal: this.props.disabled || false,
      indeterminateInternal: this.props.indeterminate || false
    };
  }

  // Here we initialize a foundation class, passing it an adapter which tells it how to
  // work with the React component in an idiomatic way.
  foundation = new MDCCheckboxFoundation({
    addClass: className => this.setState(prevState => ({
      classes: R.append(className, prevState.classes)
    })),
    removeClass: className => this.setState(prevState => ({
      classes: R.without(className, prevState.classes)
    })),
    registerAnimationEndHandler: handler => {
      if (this.refs.root) {
        this.refs.root.addEventListener(getCorrectEventName(window, 'animationend'), handler);
      }
    },
    deregisterAnimationEndHandler: handler => {
      if (this.refs.root) {
        this.refs.root.removeEventListener(getCorrectEventName(window, 'animationend'), handler)
      }
    },
    registerChangeHandler: handler => {
      // Note that this could also be handled outside of using the native DOM API.
      // For example, onChange within render could delegate to a function which calls
      // the handler passed here, as well as performs the other business logic. The point
      // being our foundations are designed to be adaptable enough to fit the needs of the host
      // platform.
      if (this.refs.nativeCb) {
        this.refs.nativeCb.addEventListener('change', handler);
      }
    },
    deregisterChangeHandler: handler => {
      if (this.refs.nativeCb) {
        this.refs.nativeCb.removeEventListener('change', handler);
      }
    },
    getNativeControl: () => {
      if (!this.refs.nativeCb) {
        throw new Error('Invalid state for operation');
      }
      return this.refs.nativeCb;
    },
    forceLayout: () => {
      if (this.refs.nativeCb) {
        this.refs.nativeCb.offsetWidth;
      }
    },
    isAttachedToDOM: () => Boolean(this.refs.nativeCb),
  });

  // For browser compatibility we extend the default adapter which checks for css variable support.
  rippleFoundation = new MDCRippleFoundation(Object.assign(MDCRipple.createAdapter(this), {
    isUnbounded: () => true,
    isSurfaceActive: () => this.refs.nativeCb[MATCHES](':active'),
    addClass: className => this.setState(prevState => ({
      classes: R.append(className, prevState.classes)
    })),
    removeClass: className => this.setState(prevState => ({
      classes: R.without(className, prevState.classes)
    })),
    registerInteractionHandler: (evtType, handler) => {
      this.refs.nativeCb.addEventListener(evtType, handler);
    },
    deregisterInteractionHandler: (evtType, handler) => {
      this.refs.nativeCb.removeEventListener(evtType, handler);
    },
    updateCssVariable: (varName, value) => {
      this.setState(prevState => ({
        rippleCss: R.assoc(varName, value, prevState.rippleCss)
      }));
    },
    computeBoundingRect: () => {
      const {left, top} = this.refs.root.getBoundingClientRect();
      const DIM = 40;
      return {
        top,
        left,
        right: left + DIM,
        bottom: top + DIM,
        width: DIM,
        height: DIM,
      };
    },
  }));

  render() {
    // Within render, we generate the html needed to render a proper MDC-Web checkbox.
    return (
      <div className={`mdc-checkbox ${this.state.classes.join(' ')}`} ref={e => this.refs.root = e}>
        <input id={this.props.id}
               type="checkbox"
               className="mdc-checkbox__native-control"
               aria-labelledby={this.props.labelId}
               checked={this.state.checkedInternal}
               disabled={this.state.disabledInternal}
               onChange={evt => {
                 this.setState({
                   checkedInternal: this.refs.nativeCb.checked,
                   indeterminateInternal: false
                 });
                 this.props.onChange(evt);
               }}
               ref={e => {
                 this.refs.nativeCb = e
               }}/>
        <div className="mdc-checkbox__background">
          <svg className="mdc-checkbox__checkmark"
               viewBox="0 0 24 24">
            <path className="mdc-checkbox__checkmark__path"
                  fill="none"
                  stroke="white"
                  d="M1.73,12.91 8.1,19.28 22.79,4.59"/>
          </svg>
          <div className="mdc-checkbox__mixedmark"></div>
        </div>
      </div>
    );
  }

  // Within the two component lifecycle methods below, we invoke the foundation's lifecycle hooks
  // so that proper work can be performed.
  componentDidMount() {
    this.foundation.init();
    this.rippleFoundation.init();
  }

  componentWillUnmount() {
    this.rippleFoundation.destroy();
    this.foundation.destroy();
  }

  // Here we synchronize the internal state of the UI component based on what the user has specified.
  componentWillReceiveProps(props) {
    if (props.checked !== this.props.checked) {
      this.setState({checkedInternal: props.checked, indeterminateInternal: false});
    }
    if (props.indeterminate !== this.props.indeterminate) {
      this.setState({indeterminateInternal: props.indeterminate});
    }
    if (props.disabled !== this.props.disabled) {
      this.setState({disabledInternal: props.disabled});
    }
  }

  // Since we cannot set an indeterminate attribute on a native checkbox, we use componentDidUpdate to update
  // the indeterminate value of the native checkbox whenever a change occurs (as opposed to doing so within
  // render()).
  componentDidUpdate() {
    if (this.refs.nativeCb) {
      this.refs.nativeCb.indeterminate = this.state.indeterminateInternal;
    }
    // To make the ripple animation work we update the css properties after React finished building the DOM.
    if (this.refs.root) {
      R.forEachObjIndexed((v, k) => {
        this.refs.root.style.setProperty(k, v);
      }, this.state.rippleCss);
    }
  }
}
```

**Comment by traviskaufman** [comment]
Even though we don't intend on supporting loading our entire library - which includes our vanilla components - from the server, enough people have raised issues around this where I feel like it's just worth moving that check into an initialization routine. WDYT @material-components/mdc-web team?

**Comment by yeelan0319** [comment]
We will move `getMatchesProperty` into `createAdapter` to prevent reference error when load component in non-HTML environments. 

The `createAdapter` method should be called in `componentDidMount` to run correctly.

**Comment by traviskaufman** [comment]
Resolved by #523
